### DISCUSSION TITLE
Authentication challenge redirection results in 404

### DISCUSSION THREAD

**Discussion by NCC1701M** [community/community#17858] [Codespaces]
I'm developing an ASP.NET Core (.Net 6) web app and I'm using Microsoft.Identity.Web to authenticate on a single tenant Azure AD.

Debugging the app in a local devcontainer everything works great. Now I was to going to try out Codespaces.

After several minor changes to the project it's basically running except for the Authentication.

To force the user to authenticate I use the following code snippet

```csharp
var scheme = OpenIdConnectDefaults.AuthenticationScheme;
var redirectUrl = ctx.Request.Path.Value.ToLower();
await ctx.ChallengeAsync(
	scheme,
	new AuthenticationProperties { RedirectUri = redirectUrl }
);
```

After the user was able to authenticate,  instead of redirecting to `https://myuser-myproject-someid-1234.githubpreview.dev` the lib redirects to `http://localhost` which of course won't work.

So after some research I found a solution to tell the identity lib to use the correct codespace url

```csharp
#if DEBUG
services.Configure<OpenIdConnectOptions>(OpenIdConnectDefaults.AuthenticationScheme, options =>
{
	var redirectHandler = options.Events.OnRedirectToIdentityProvider;
	options.Events.OnRedirectToIdentityProvider = async ctx =>
	{
		await redirectHandler(ctx);
		// The original value of ctx.ProtocolMessage.RedirectUri is https://localhost/signin-oidc
		ctx.ProtocolMessage.RedirectUri = "https://myuser-myproject-someid-1234.githubpreview.dev/signin-oidc";
	};
});
#endif
```

Now it's going to redirect to the correct url but the browser can't find the address and shows the following error
![grafik](https://user-images.githubusercontent.com/7260053/171848409-89ae04f0-6316-483e-8b17-e1d56946fae4.png)

So how can I solve this problem?

**Comment by aditya30394** [comment]
Hey @NCC1701M,
Thanks for trying codespaces and reaching out to us.
We would need a bit more information to help you out and reproduce this issue on our end. A bare bone repo would also be fine. 

A couple of observations:

1. You should not be required to adjust the localhost url to a "githubpreview.dev" url for your app to work. You mentioned that you were able to run your app in local devcontainer. So, if locally you were running on, say, port 5000 you would be using (locally) a URL like "http://localhost:5000" - assuming that your app is configured to run on port 5000. Once you start working on codespaces:
  a. If you are using [github codespaces extension via vscode](https://marketplace.visualstudio.com/items?itemName=GitHub.codespaces), then just run the app and forward the port 5000 (by going to ports view). You can then access your app by using the localhost url -> http://localhost:5000
  b. If you are developing in web browser then after forwarding a port you would be shown a non-localhost url. If you forwarded port 5000 then the URL would look like "codespacename-5000.githubpreview.dev". Note that at no point you need to make changes to your code for this to work.
  2. I created this bare minimum webapp - https://github.com/aditya30394/MyWebAppTest/tree/main for reference. I can run this locally by doing "dotnet watch" and accessing the app at "(http://localhost:5081" (my app is configured to run at 5001). To use this app in codespaces, I don't need to change anything. I can still do "dotnet watch" and access my app using the URL shown in ports tab

Here's a small demo. Hopefully this helps:

![customer_web_app](https://user-images.githubusercontent.com/8336542/172734722-dc0edf4d-d385-43c2-b964-6fae8ab67390.gif)

**Comment by NCC1701M** [comment]
Hi @aditya30394 

thanks for your response. I've set up a little Repo with nearly the same configuration I used to run the real project:
https://github.com/NCC1701M/AuthIssueTest

I use the 5001 port and HTTPS. I'm always open the app by clicking the entry in the *Ports* tab.

When I'm running the app in a local devcontainer it works as you can see in the follwing gif:
![redirect_localhost_success](https://user-images.githubusercontent.com/7260053/173032957-5aa7a6e4-5238-48eb-a110-df4d87ac7709.gif)


When I'm running the same app in codespaces this is what happens:
![redirect_localhost_fail](https://user-images.githubusercontent.com/7260053/173033426-4ec9a55f-765a-4819-ab04-16bc5f80f972.gif)

And when I use the snippet

```csharp
#if DEBUG
services.Configure<OpenIdConnectOptions>(OpenIdConnectDefaults.AuthenticationScheme, options =>
{
	var redirectHandler = options.Events.OnRedirectToIdentityProvider;
	options.Events.OnRedirectToIdentityProvider = async ctx =>
	{
		await redirectHandler(ctx);
		// The original value of ctx.ProtocolMessage.RedirectUri is https://localhost/signin-oidc
		ctx.ProtocolMessage.RedirectUri = "https://myuser-myproject-someid-1234.githubpreview.dev/signin-oidc";
	};
});
#endif
```

This is what happens (the redirect to login.microsoftonline.com is to fast to see but it's there):
![redirect_githubdev_fail](https://user-images.githubusercontent.com/7260053/173033636-6ad52f75-9e4a-43df-9341-2e80c33d3479.gif)

**Comment by aditya30394** [comment]
Hey @NCC1701M,
Thanks for sharing the repo for reproduction of the issue. We were able to reproduce it on our end. Just to confirm, you had to configure your app registration to allow the *.githubpreview.dev redirection, correct? SOmething like this:

![image](https://user-images.githubusercontent.com/8336542/173162651-b4bdd088-d2c3-4a2a-85ed-98dab1d2be0b.png)


While we investigate the root cause for not being able to use the URLs, we recommend that you make use of our [Github codesapces](https://marketplace.visualstudio.com/items?itemName=GitHub.codespaces) extension on vscode IDE. 

I was able to run the app without making any changes with the extension. That would allow you to continue using codespaces for your development - something that we are extremely glad that you are doing. Please continue that. 

From the configuration, I can see that you are trying to use http protocol on port 5000 and https on 5001. By default the extension will forward the 5001 port over http. So, if you want to make use of https protocol, we recommend that you disable the "Use port forwarding service for HTTPS" setting in extension.
![image](https://user-images.githubusercontent.com/8336542/173163033-0145c56c-1117-45fe-9db0-7c6caadeccae.png)

**Reply by NCC1701M** [reply]
Hi @aditya30394 

Thank you for your reply. I will test the extension on monday and give you a response.

> Just to confirm, you had to configure your app registration to allow the *.githubpreview.dev redirection, correct?

That is correct.

**Comment by aditya30394** [comment]
Hey @NCC1701M,
Hopefully it works with desktop extension. 
Can you also try renaming "signin" to something else like "login" in the redirect URL? We believe that the "/signin" path is conflicting with the similar endpoint in our service and would like to confirm our theory. So, it will be great if you can test replacing "/signin-oidc" with "login-oidc". 

Also, it will be great if you can share the name of your codespace with us - it is the first part of URL. URLs have format of "name-port.githubpreview.dev". We can look up our telemetry to see where's the issue happening in service.

**Comment by NCC1701M** [comment]
Hi @aditya30394  ,

I've renamed signin-oidc with login-oidc. It works in combination with the `ctx.ProtocolMessage.RedirectUri = "https://myuser-myproject-someid-1234.githubpreview.dev/signin-oidc";` part. Without it still redirects to `http://localhost/login-oidc`.

When the redirection was correct it loads the `index.html` provided by the `UseProxyToSpaDevelopmentServer` but for all other files provided by the spa proxy (.js, .css, .ico) I receive a `net::ERR_HTTP2_PROTOCOL_ERROR` error.

![image](https://user-images.githubusercontent.com/7260053/173583705-486e5113-ce38-4808-8c22-b978ff8a663e.png)


Unfortunatly I had only the chance to try the Codespaces extension with the disabled `Use port forwarding service for HTTPS` a few minutes and it seems to be working. But I will do some more tests soon and let you know.

The name of the codespace I've did the last tests with is: ncc1701m-authissuetest-r4wx7w9vhwpv7
